File: task_test.mjs
// npm i protobufjs --save
import protobuf from 'protobufjs'
import {Task} from './task/task.pb.mjs'
import Test from './test.mjs'

const r = 1;

(() => {
  const t = new Test(async () => {
    return JSON.stringify({text: 'test', done: true})
  }, 'json encode')
  t.bench(r)
  t.test(JSON.stringify({text: 'test', done: true}))
})();

(() => {
  const b = JSON.stringify({text: 'test', done: true})
  const t = new Test(async () => {
    return JSON.parse(b)
  }, 'json decode')
  t.bench(r)
  t.test({text: 'test', done: true})
})();

(() => {
  const t = new Test(async () => {
    const msg = Task.create({text: 'test', done: true})
    return Task.encode(msg).finish()
  }, 'pbf encode')
  t.bench(r)
  t.test({text: 'test', done: true}) // TODO
})();

(() => {
  const msg = Task.create({text: 'test', done: true})
  const bin = Task.encode(msg).finish()
  const t = new Test(async () => {
    return Task.decode(bin)
  }, 'pbf decode')
  t.bench(r)
  t.test({text: 'test', done: true}) // TODO
})();

protobuf.load('./task/task.proto').then((root) => {
  const Task = root.lookup('Task')

  (() => {
    const t = new Test(async () => {
      const writer = protobuf.Writer.create() // new protobuf.BufferWriter()
      Task.encodeDelimited({text: 'test', done: true}, writer)
      writer.finish()
      return
    }, 'pbf encode delimited')
    t.bench(r)
    t.test({text: 'test', done: true}) // TODO
  })();

  (() => {
    const writer = protobuf.Writer.create() // new protobuf.BufferWriter()
    Task.encodeDelimited({text: 'test', done: true}, writer)
    writer.finish()
    const t = new Test(async () => {
      const result = []
      const reader = protobuf.Reader.create(writer)
      while (reader.pos < reader.len) {
        result.push(Task.decodeDelimited(reader))
      }
      return result
    }, 'pbf decode delimited')
    t.bench(r)
    t.test([{text: 'test', done: true}])
  })();
});

File: task_test.go
package main

import (
	"encoding/json"
	testing "testing"

	"./task"
	proto2 "github.com/gogo/protobuf/proto"
	"github.com/golang/protobuf/proto"
)

// const UINT64 = uint64(unsafe.Sizeof(uint64(0)))

//go:generate bash -c ./proto.sh

func BenchmarkMarshalJson(b *testing.B) {
	msg := &task.Task{
		Text: "test",
		Done: false,
	}

	for n := 0; n < b.N; n++ {
		_, err := json.Marshal(msg)
		if err != nil {
			b.Error(err)
		}
	}
}

func BenchmarkUnMarshalJson(b *testing.B) {
	msg := &task.Task{
		Text: "test",
		Done: false,
	}

	bin, err := json.Marshal(msg)
	if err != nil {
		b.Error(err)
		return
	}

	var task task.Task
	for n := 0; n < b.N; n++ {
		err := json.Unmarshal(bin, &task)
		if err != nil {
			b.Error(err)
			return
		}
	}
}

func BenchmarkMarshalProto(b *testing.B) {
	msg := &task.Task{
		Text: "test",
		Done: false,
	}

	for n := 0; n < b.N; n++ {
		_, err := proto.Marshal(msg)
		if err != nil {
			b.Error(err)
		}
	}
}

func BenchmarkUnMarshalProto(b *testing.B) {
	msg := &task.Task{
		Text: "test",
		Done: false,
	}

	bin, err := proto.Marshal(msg)
	if err != nil {
		b.Error(err)
		return
	}

	var t task.Task
	for n := 0; n < b.N; n++ {
		err := proto.Unmarshal(bin, &t)
		if err != nil {
			b.Error(err)
			return
		}
	}
}

func BenchmarkMarshalProto2(b *testing.B) {
	msg := &task.Task{
		Text: "test",
		Done: false,
	}

	for n := 0; n < b.N; n++ {
		_, err := proto2.Marshal(msg)
		if err != nil {
			b.Error(err)
		}
	}
}

func BenchmarkUnMarshalProto2(b *testing.B) {
	msg := &task.Task{
		Text: "test",
		Done: false,
	}

	bin, err := proto2.Marshal(msg)
	if err != nil {
		b.Error(err)
		return
	}

	var t task.Task
	for n := 0; n < b.N; n++ {
		err := proto2.Unmarshal(bin, &t)
		if err != nil {
			b.Error(err)
			return
		}
	}
}
